src/diffinterface.jl

```diff
@@ -0,0 +1,393 @@
+end
+
+
+
+@inline tupleconcat(x) = x
+@inline tupleconcat(x, y) = (x..., y...)
+@inline tupleconcat(x, y, z...) = (x..., tupleconcat(y, z...)...)
+
+@inline function derivative(mode::ForwardMode, f, x; shadow=onehot(x))
+    values(only(autodiff(mode, f, BatchDuplicatedNoNeed, BatchDuplicated(x, shadow))))
+end
+@inline function derivative_deferred(mode::ForwardMode, f, x; shadow=onehot(x))
+    values(only(autodiff_deferred(mode, f, BatchDuplicatedNoNeed, BatchDuplicated(x, shadow))))
+end
+
+@inline function derivative(mode::ForwardMode, f::F, x::X, ::Val{chunk};
+                            shadow=chunkedonehot(x, Val(chunk))) where {F,X,chunk}
+    if chunk == 0
+        throw(ErrorException("Cannot differentiate with a batch size of 0"))
+    end
+    tmp = ntuple(length(shadow)) do i
+        values(autodiff(mode, f, BatchDuplicatedNoNeed, BatchDuplicated(x, shadow[i]))[1])
+    end
+    tupleconcat(tmp...)
+end
+
+@inline function valderivative(mode::ForwardMode, f, x; shadow=onehot(x))
+    (y, df) = autodiff(mode, f, BatchDuplicated, BatchDuplicated(x, shadow))
+    (y, values(df))
+end
+
+@inline function _split_tuple_of_tuples(tpl::Tuple)
+    t1 = ntuple(i -> tpl[i][1], length(tpl))
+    t2 = ntuple(i -> values(tpl[i][2]), length(tpl))        
+    (t1, tupleconcat(t2...))
+end
+
+# note that we don't provide valderivative for batched mode because it causes it to re-evaluate
+
+@inline gradient_output(df, x) = df
+
+# this should handle general mutable array types
+@inline gradient_output(df, x::AbstractArray) = copyto!(similar(x), df)
+
+@inline function gradient(mode::ForwardMode, f::F, x::X; shadow=onehot(x)) where {F,X}
+    df = derivative(mode, f, x; shadow)
+    gradient_output(df, x)
+end
+
+@inline function gradient(mode::ForwardMode, f::F, x::X, ::Val{chunk};
+                          shadow=chunkedonehot(x, Val(chunk))) where {F,X,chunk}
+    df = derivative(mode, f, x, Val(chunk); shadow)
+    gradient_output(df, x)
+end
+
+@inline function valgradient(mode::ForwardMode, f::F, x::X; shadow=onehot(x)) where {F,X}
+    (y, df) = valderivative(mode, f, x; shadow)
+    (y, gradient_output(df, x))
+end
+
+#WARN: presumably this is done in this particular way for efficiency, but it's not super clear that
+# it'll always give consistent results...
+# Before this PR, this would return type X even though forward mode gradients always returned tuple
+# I think it's consistent now?
+@inline function gradient(mode::ReverseMode, f::F, x::X) where {F,X}
+    if Compiler.active_reg_inner(X, #=seen=#(), #=world=#nothing, #=justActive=#Val(true)) == Compiler.ActiveState
+        dx = Ref(make_zero(x))
+        autodiff(mode, f, Active, MixedDuplicated(x, dx))
+        return only(dx)
+    else
+        dx = make_zero(x)
+        autodiff(mode, f, Active, Duplicated(x, dx))
+        return dx
+    end
+end
+
+"""
+    gradient_deferred(::ReverseMode, f, x)
+
+Like [`gradient`](@ref), except it using deferred mode.
+"""
```

Comment:
I tried adding this earlier but @vchuravy objected. You'll have to convince him on this one

Reply:
Is the issue that `deferred` is eventually expected to be deprecated so you don't want more API surface that refers to it?